unidist/core/backends/mpi/core/communication.py

```diff
@@ -191,27 +246,78 @@ def send_complex_data(comm, data, dest_rank):
         Data object to send.
     dest_rank : int
         Target MPI process to transfer data.
+
+    Returns
+    -------
+    object
+        A serialized msgpack data.
+    list
+        A list of pickle buffers.
+    list
+        A list of buffers amount for each object.
     """
-    serializer = MPISerializer()
+    serializer = ComplexSerializer()
     # Main job
     s_data = serializer.serialize(data)
-
-    # Send message pack bytestring
-    comm.send(len(s_data), dest=dest_rank)
-    comm.Send([s_data, MPI.CHAR], dest=dest_rank)
     # Retrive the metadata
     raw_buffers = serializer.buffers
     len_buffers = serializer.len_buffers
 
+    # MPI comminucation
+    _send_complex_data_impl(comm, s_data, raw_buffers, len_buffers, dest_rank)
+
+    # For caching purpose
+    return s_data, raw_buffers, len_buffers
+
+
+def _isend_complex_data_impl(comm, s_data, raw_buffers, len_buffers, dest_rank):
+    """
+    Send serialized complex data.
+
+    Non-blocking asynchronous interface.
+
+    Parameters
+    ----------
+    comm : object
+        MPI communicator object.
+    s_data: object
+        A serialized msgpack data.
+    raw_buffers : list
+        A list of pickle buffers.
+    len_buffers : list
+        A list of buffers amount for each object.
+    dest_rank : int
+        Target MPI process to transfer data.
+
+    Returns
+    -------
+    list
+        A list of pairs, ``MPI_Isend`` handler and associated data to send.
+    """
+    handlers = []
+
+    # Send message pack bytestring
+    h1 = mpi_isend_object(comm, len(s_data), dest_rank)
+    h2 = mpi_isend_buffer(comm, s_data, dest_rank)
```

Comment:
We were already discussing this and decided to keep MPI-like interfaces, i.e. with/without `i` prefix.

Reply:
What's advantage from this?

Reply:
Consistent API between MPI and unidist.MPI interfaces. Also, such unification do not lead to the lines of code reduction.

Reply:
What's need we have in consistent API between MPI and unidist.MPI?

One function have a number advantages:

1. This one function really reduces amount of the code (1 API instead of 4). Besides, this func can be run only one time for sending object with followed buffer (a lot of places where you can see this pair). Example:
instead of this
```python
h1 = mpi_isend_object(comm, len(s_data), dest_rank)
h2 = mpi_isend_buffer(comm, s_data, dest_rank)
```
just
```python
h1, h2 = mpi_send(comm, len(s_data), s_data, dest_rank)
```
2. 4 funcs overload the implementation, one entry point for some functionality is a better solution in this case.  
3. One function with one description of approaches of buffer sending is more convenient instead of 4 description which are placed in different parts of the file.

Reply:
I agree that send_buffer can be combined with sending length and buffer data itself, but the function that returns handler in one case and returns nothing in another is confusing.

Reply:
> Also, by "not reduce amount of code" I mean no code reuse happens here, instead of 2 functions (async and not async)

*Instead of 4 functions https://github.com/modin-project/unidist/pull/53#discussion_r787718816.

But this dispatching is placed in one place which will be more easy to understand by future contributors.

If double-type return is confusing, you can set a handler to `None` in a case of sync execution

Reply:
You have a point there, I agree. But proposed approach contradicts with a KISS principle. Is this comment blocks the PR to move on?

Reply:
I don't think, that proposed func above is too difficult to contradicts with the principle. I don't mind if this change will be done as a separate issue. In this case, please, create the issue.

Reply:
#100 issue is opened for this purpose